oscaralanpierce/skyrim_inventory_management_frontend
# Refresh token and retry when API returns a 401 response - take 2

## Context

[**Enable the back end to refresh tokens**](https://trello.com/c/V087rV56/290-enable-the-back-end-to-refresh-tokens)

In #68, we attempted to catch 401 responses from the API and retry the requests after refreshing the Google access token. Unfortunately, this approach did not work during manual testing. This PR is take 2.

## Changes

* Ensure Google access tokens are refreshed and API calls retried if the API returns a 401 response
* Remove `requireLogin` function from `LoginProvider` and remove from `useEffect` hooks in `ShoppingListsProvider` and `GamesProvider`
* Run formatter (hence random changed ADR)

## Required Tasks


## Considerations

There's much about this approach I'm not happy with, starting with the two additional positional args every function that makes an API call now has to take (`idToken` and `retries`). I wanted to give these args default values to make them a little less onerous, but that would involve putting them before the optional args (mostly `onSuccess` and `onError` callbacks), forcing a value to be included every time the functions were called with optional args. I wanted these values to be "behind the scenes" and not that visible to any old context consumer.

## Manual Test Cases

Manually testing this PR will be a slow process since we'll have to manually test each function after waiting for the access token to expire. It is also completely essential since automated testing was impractical. The following functions will need to be tested, after waiting for the token to expire and without refreshing the page:

- [x] Create a game from the games page
- [x] Update a game on the games page
- [x] Destroy a game on the games page
- [x] Fetch shopping lists for another game on the shopping lists page (just change game in dropdown)
- [x] Create a shopping list on the shopping lists page
- [x] Update the title of a shopping list on the shopping lists page
- [x] Destroy a shopping list on the shopping lists page
- [x] Create a shopping list item on the shopping lists page
- [x] Update a shopping list item on the shopping lists page
- [x] Destroy a shopping list item on the shopping lists page

Remember that updating code refreshes the page, so we can't make code changes while waiting. Additionally, fetching games also now has the same retry behaviour, however, it is difficult to effect a fetch of games without refreshing the page.

To verify the functionality is working, after performing each action, look at the back end logs to ensure two requests have been made. The first should've resulted in a 401 response, and the second should be successful.
